### PR TITLE
Allow for transforms on the model before MLFlow registration

### PR DESCRIPTION
**Pull request by snarayan21**
Allows subclasses of HF Checkpointer to transform the model before MLFlow registration, if they so choose. Also slight change to make `using_peft` an attribute of HF Checkpointer instance, allowing it to be used / set.

### CODE REVIEW

File: llmfoundry/callbacks/hf_checkpointer.py [hunk 1]
```diff
@@ -480,22 +500,20 @@ def dtensor_to_tensor_hook(
 
             log.debug(f'Creating new model instance')
 
-            if composer_model.using_peft:
-                # We don't use meta here because the state dict does not contain the full
```
Comment (snarayan21) [inline llmfoundry/callbacks/hf_checkpointer.py:484]: want to point out that this comment is wrong -- the state dict does contain the full model in addition to the adapter weights.

This is important & needed if we want to merge the adapters later on. So it's a good thing lol

**Review by snarayan21 — COMMENTED** [review]
note

File: llmfoundry/callbacks/hf_checkpointer.py [hunk 2]
```diff
@@ -406,6 +424,8 @@ def _save_checkpoint(self, state: State, logger: Logger):
             state_dict_model = state.model.model
             original_tokenizer = state.model.tokenizer
 
+        self.using_peft = composer_model.using_peft
+
         if version.parse(torch.__version__) > version.parse('2.2.9'):
```
[the remark this answers is not in the document]
Comment (snarayan21) [reply llmfoundry/callbacks/hf_checkpointer.py:429]: yep

File: llmfoundry/callbacks/hf_checkpointer.py [hunk 3]
```diff
@@ -480,22 +500,20 @@ def dtensor_to_tensor_hook(
...
+                if self.using_peft:
+                    active_adapter = original_model.active_adapter
+                    base_model = original_model.get_base_model()
+                    new_base_model_instance = type(base_model)(new_config)
+
+                    new_model_instance = type(original_model)(
+                        new_base_model_instance,
+                        original_model.peft_config[active_adapter],
+                    )
+                    new_model_instance.to(dtype=self.dtype)
```
[the remark this answers is not in the document]
Comment (snarayan21) [reply llmfoundry/callbacks/hf_checkpointer.py:515]: true

File: llmfoundry/callbacks/hf_checkpointer.py [hunk 4]
```diff
@@ -406,6 +424,8 @@ def _save_checkpoint(self, state: State, logger: Logger):
             state_dict_model = state.model.model
             original_tokenizer = state.model.tokenizer
 
+        self.using_peft = composer_model.using_peft
```
[the remark this answers is not in the document]
Comment (snarayan21) [reply llmfoundry/callbacks/hf_checkpointer.py:427]: done

File: llmfoundry/callbacks/hf_checkpointer.py [hunk 5]
```diff
@@ -556,6 +574,11 @@ def dtensor_to_tensor_hook(
 
         if dist.get_global_rank() == 0:
             if self.mlflow_registered_model_name and self._is_last_batch(state):
+
+                new_model_instance = self.transform_model_pre_registration(
```
[the remark this answers is not in the document]
Comment (snarayan21) [reply llmfoundry/callbacks/hf_checkpointer.py:581]: done

File: llmfoundry/callbacks/hf_checkpointer.py [hunk 6]
```diff
@@ -274,6 +279,15 @@ def run_event(self, event: Event, state: State, logger: Logger) -> None:
                 mlflow.environment_variables.MLFLOW_HUGGINGFACE_MODEL_MAX_SHARD_SIZE.set(
                     '1GB',
                 )
+
+            # Check if the model is using PEFT
+            if state.is_model_ddp:
+                composer_model = state.model.module
+            elif isinstance(state.model.model, FSDP):
```
Comment (dakinggg) [inline llmfoundry/callbacks/hf_checkpointer.py:286]: don't need three branches here, just two

File: llmfoundry/callbacks/hf_checkpointer.py [hunk 7]
```diff
@@ -362,6 +376,23 @@ def transform_config(
                 copied_config.ffn_config['moe_world_size'] = 1
         return copied_config
 
+    def transform_model_pre_registration(
+        self,
+        model: PreTrainedModel,
+    ) -> PreTrainedModel:
+        """Transform the model before registering with MLFlow.
```
Comment (dakinggg) [inline llmfoundry/callbacks/hf_checkpointer.py:383]:
```suggestion
        """Transform the model before registering with MLflow.
```

File: llmfoundry/callbacks/hf_checkpointer.py [hunk 8]
```diff
@@ -362,6 +376,23 @@ def transform_config(
...
+
+        This allows a subclass to modify the model before registering with MLFlow. The base class implementation will
```
Comment (dakinggg) [inline llmfoundry/callbacks/hf_checkpointer.py:385]:
```suggestion
        This allows a subclass to modify the model before registering with MLflow. The base class implementation will
```

**Review by dakinggg — APPROVED** [review]

File: llmfoundry/callbacks/hf_checkpointer.py (hunk 6 above)
Comment (snarayan21) [reply llmfoundry/callbacks/hf_checkpointer.py:286]: woops

**Pull request merged**